libsplinter/src/rest_api/auth/allow_keys.rs

```diff
@@ -0,0 +1,401 @@
+            _ => Ok(AuthorizationHandlerResult::Continue),
+        }
+    }
+
+    fn clone_box(&self) -> Box<dyn AuthorizationHandler> {
+        Box::new(self.clone())
+    }
+}
+
+/// Internal state of the authorization handler
+struct Internal {
+    file_path: String,
+    cached_keys: Vec<PublicKey>,
+    last_read: SystemTime,
+}
+
+impl Internal {
+    fn new(file_path: &str) -> Result<Self, InternalError> {
+        let mut internal = Self {
+            file_path: file_path.into(),
+            cached_keys: vec![],
+            last_read: SystemTime::UNIX_EPOCH,
+        };
+
+        // If file already exists, read it; otherwise create it.
+        if PathBuf::from(file_path).is_file() {
+            internal.read_keys()?;
+        } else {
+            File::create(file_path).map_err(|err| {
+                InternalError::from_source_with_message(
+                    Box::new(err),
+                    "failed to create allow keys file".into(),
+                )
+            })?;
+            internal.last_read = SystemTime::now();
+        }
+
+        Ok(internal)
+    }
+
+    /// Get the internal list of keys. If the backing file has been modified since the last read,
+    /// attempt to refresh the cache.
+    fn get_keys(&mut self) -> Vec<PublicKey> {
+        let file_read_result = std::fs::metadata(&self.file_path)
+            .and_then(|metadata| metadata.modified())
+            .map_err(|err| {
+                InternalError::from_source_with_message(
+                    Box::new(err),
+                    "failed to read allow keys file's last modification time".into(),
+                )
+            })
+            .and_then(|last_modified| {
+                if last_modified > self.last_read {
+                    self.read_keys()
+                } else {
+                    Ok(())
+                }
+            });
+
+        // Log any errors that occurred with checking or reading the backing file and use the
+        // in-memory cache.
+        if let Err(err) = file_read_result {
+            warn!(
+                "Using cached keys; failed to read from allow keys file: {}",
+                err
+            );
+        }
+
+        self.cached_keys.clone()
+    }
+
+    /// Reads the backing file and caches its contents, logging an error for any key that can't be
+    /// read
+    fn read_keys(&mut self) -> Result<(), InternalError> {
+        let file = File::open(&self.file_path).map_err(|err| {
+            InternalError::from_source_with_message(
+                Box::new(err),
+                "failed to open allow keys file".into(),
+            )
+        })?;
+        let keys = BufReader::new(file)
+            .lines()
+            .filter_map(|res| match res {
+                Ok(line) => match PublicKey::new_from_hex(&line) {
+                    Ok(key) => Some(key),
+                    Err(err) => {
+                        error!("Invalid key in allow keys file: {}", err);
```

Comment:
If you were to enumerate the iterator, you could log which line the bad key was on (the err probably won't give any details about which key it was, just what was wrong with the key)

Reply:
Good idea; updated.